src/tir/transforms/common_subexpr_elim_tools.h

```diff
@@ -0,0 +1,216 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file common_subexpr_elim_tools.h
+ * \brief Interface of analysis tools and utility functions used
+           by the Common Subexpression Elimination (CSE) pass.
+ */
+
+#ifndef TVM_TIR_TRANSFORMS_COMMON_SUBEXPR_ELIM_TOOLS_H_
+#define TVM_TIR_TRANSFORMS_COMMON_SUBEXPR_ELIM_TOOLS_H_
+
+#include <tvm/runtime/container/string.h>
+#include <tvm/tir/analysis.h>  // For the ExprDeepEqual analysis
+#include <tvm/tir/expr.h>
+#include <tvm/tir/expr_functor.h>
+#include <tvm/tir/stmt.h>
+#include <tvm/tir/stmt_functor.h>  // For the class StmtExprVisitor
+
+#include <unordered_map>  // For the hashtable datatype
+#include <vector>
+
+#include "../../../3rdparty/dmlc-core/include/dmlc/optional.h"
+
+namespace tvm {
+namespace tir {
+
+/*!
+ * \brief A table of computations is a hashtable which associates to each expression being computed
+          a number (which is the number of time that it is computed)
+					It is important to note that the hash used is a StructuralHash (and not an ObjectPtrHash)
+					as we need to hash similarly deeply equal terms.
+					The comparison used is ExprDeepEqual, which is stricter than StructuralEqual (as it does
+					not do variables remapping), so it is compatible with StructuralHash (intended to be used
+					with StructuralEqual).
+ */
+using TableOfComputations = std::unordered_map<PrimExpr, size_t, StructuralHash, ExprDeepEqual>;
+
+/*!
+ * \brief A cache of computations is made of a pair of two hashtables, which respectively associate
+          to each statement or expression of the program its table of computations. Its purpose is
+          to avoid the CSE pass from recomputing repeatedly the same tables of computations.
+ */
+struct CacheOfComputations {
+  // Part of the cache for statements
+  // It maps each known statement to its table of computations
+  std::unordered_map<Stmt, TableOfComputations, ObjectPtrHash, ObjectPtrEqual>
+      cache_stmt_table_computations_;
+
+  // Part of the cache for expressions
+  // It maps each known expression to its table of computations
+  std::unordered_map<PrimExpr, TableOfComputations, ObjectPtrHash, ObjectPtrEqual>
+      cache_expr_table_computations_;
+};
+
+/*!
+ * \brief Visitor which returns in a hashtable the (syntatic) computations done by an expression
+          or by a statement.
+ * \note Computations here are considered syntactically, meaning that semantically equivalent
+          computations that are not syntactically the same are not merged together.
+ */
+class ComputationsDoneBy : public StmtExprVisitor {
+ public:
+  // Toplevel (static) methods
+  static TableOfComputations GetComputationsDoneBy(
+      const PrimExpr& expr, std::function<bool(const PrimExpr&)> is_eligible_computation,
+      std::function<bool(const PrimExpr&)> can_contain_computations);
+  static TableOfComputations GetComputationsDoneBy(
```

Comment:
I prefer the name `ExtractSubComputations`

Reply:
While I agree that `ExtractSubComputations` is a nicer name, I think it would be better to stick with the concept of "computations being done by a node", which I have tried to describe [at the beginning of the implementation of this class](https://github.com/apache/tvm/pull/9482/files#diff-f83c46530e92c628fd309499ceffa32cb2fb0505633ac0e754e2bdef4d518962R68-R84).

I agree that `ExtractSubComputations` is a nicer name, but it would lie quite a bit about what the function really is returning. For an expression in input, it indeed returns its subexpressions (if recursing inside the current node is allowed by the second predicate `can_contain_computations`), but that's only if the original expression itself was not illegible (according to the first predicate `is_eligible_computation`), as the original expression is precisely what would have been returned otherwise.